lib/Arch/Sleigh/PPCArch.cpp

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright (c) 2022-present Trail of Bits, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "PPC.h"
+
+#define INCLUDED_FROM_REMILL
+#include <remill/Arch/PPC/Runtime/State.h>
+
+namespace remill {
+
+namespace sleighppc {
+
+static const char *kPPCVLERegName = "VLEReg";
+
+SleighPPCDecoder::SleighPPCDecoder(const remill::Arch &arch)
+    : SleighDecoder(arch, "ppc_64_isa_vle_be.sla", "ppc_64.pspec",
+                    {{"vle", kPPCVLERegName}}, {}) {}
+
+llvm::Value *SleighPPCDecoder::LiftPcFromCurrPc(llvm::IRBuilder<> &bldr,
+                                                llvm::Value *curr_pc,
+                                                size_t curr_insn_size) const {
+  // PC on thumb points to the next instructions next.
+  return bldr.CreateAdd(
+      curr_pc, llvm::ConstantInt::get(curr_pc->getType(), curr_insn_size));
+}
+
+void SleighPPCDecoder::InitializeSleighContext(
+    remill::sleigh::SingleInstructionSleighContext &ctxt,
+    const std::map<std::string, uint64_t> &context_values) const {
+  // If the context value mappings specify a value for the VLE register, let's pass that into
+  // Sleigh.
+  //
+  // Otherwise, default to VLE off.
+  uint64_t vle_value = 0;
+  const auto vle_iter = context_values.find(kPPCVLERegName);
+  if (vle_iter != context_values.end()) {
+    vle_value = vle_iter->second;
+  }
+  ctxt.GetContext().setVariableDefault("vle", vle_value);
+}
+
+class SleighPPCArch : public ArchBase {
+ public:
+  SleighPPCArch(llvm::LLVMContext *context_, OSName os_name_,
+                ArchName arch_name_)
+      : ArchBase(context_, os_name_, arch_name_),
+        decoder(*this) {}
+  virtual ~SleighPPCArch() = default;
+
+  DecodingContext CreateInitialContext(void) const override {
+    return DecodingContext().PutContextReg(kPPCVLERegName, 0);
```

Comment:
is there a way to specify which one to use?
something isn't quite right here since the tests still pass with this (they currently only test with VLE) so we're not setting this consistently everywhere it needs to

Reply:
oops, nevermind
had a stale branch checked out

this does make the tests fail (correctly) but now there's not external way (as far as i can tell) to set this to VLE mode

Reply:
Ah good point. I'll look into this.

I did it like this because Ghidra specifies a register value for `vle` when VLE mode is on but assumes it's off.

Reply:
this might be something that has to be set in the test harness actually
this is the way it's being set in the harness
```cpp
std::optional<std::pair<llvm::Function *, remill::Instruction>>
LiftingTester::LiftInstructionFunction(std::string_view fname,
                                       std::string_view bytes,
                                       uint64_t address) {
  remill::Instruction insn;
  // This works for now since each arch has an initial context that represents the arch correctly.
  if (!this->arch->DecodeInstruction(address, bytes, insn,
                                     this->arch->CreateInitialContext())) {
    LOG(ERROR) << "Failed decode";
    return std::nullopt;
  }

```

Reply:
pushed a fix for the tests, i think `CreateInitialContext` defaulting to 0 is fine